Save Search Improvements (CRASM-3418/3429)

hawkishpolicy opened the pull request:
<!-- GitHub renders PRs such that soft line breaks are treated as hard
line breaks.  In order to make this template render as expected we
therefore have to avoid soft breaks and therefore will offend
markdownlint with long lines.  This is the reason for the markdownline
disable directive just below.

For more details see:
https://github.com/github/markup/issues/1050#issuecomment-294654762 -->
<!-- markdownlint-disable MD013 -->
# <!-- Use the title to describe PR changes in the imperative mood --> #
**Improve Save Search Experience**
## 🗣 Description ##
- Imported useAreFiltersDefault hook and initialFilters prop into SaveSearchModal.
- disabled prop of modal button is now based on boolean returned from hook:
   - if filters are default, the button is disabled.
   - if filters are not, the button is enabled.
   - was originally based of filters array length being greater than 1
- Restored ability to use spaces in Save Search name.
- Added the following validation to both create_saved_search and update_saved_search calls:
   - prevent unauthorized region/org filter injection by a standard user.
   - prevent standard user from saving their default filters.
   - prevent admin users from saving their default filters.
- Refactored saved search name validation's scope to user level.
   - Was previously scoped to all users.
   - This was preventing saving of searches with common names.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##
- Allows admin users to save searches based off single region or organization.
- Prevents standard users from saving searches just based off their region and/or org.
- Provide validation to prevent filter injection.
- Provide validation to prevent saving of searches with default filters.
- Bug fixes and improvements.
- Closes CRASM-3418
- Closes CRASM-3429
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##
- tested locally
- created frontend unit test for new functionality.
- created pytest for api changes.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your code, etc. -->


## 📷 Screenshots (if appropriate) ##

**Global Admins:**
<img width="2032" height="1167" alt="Screenshot 2025-12-03 at 11 37 54 AM" src="https://github.com/user-attachments/assets/218c66d4-84f0-48c4-b0e6-6ad94bde7586" />
<i mg width="2032" height="1167" alt="Screenshot 2025-12-03 at 11 38 01 AM" src="https://github.com/user-attachments/assets/5b7d4a4d-1b4f-4110-9779-71df75f17693" />
<img width="2032" height="1167" alt="Screenshot 2025-12-03 at 11 38 14 AM" src="https://github.com/user-attachments/assets/f2bf31c3-08fd-4c4e-8df2-c5f12604375f" />
<img width="2032" height="1167" alt="Screenshot 2025-12-03 at 11 44 39 AM" src="https://github.com/user-attachments/assets/97abd10e-cf9a-4dbe-968b-907a475fd3cb" />

**Standard User:**
<img width="2032" height="1167" alt="Screenshot 2025-12-03 at 11 58 03 AM" src="https://github.com/user-attachments/assets/ed47bb75-4f3e-4242-afaf-bef75a001800" />
<img width="2032" height="1167" alt="Screenshot 2025-12-03 at 11 58 12 AM" src="https://github.com/user-attachments/assets/c612e0f6-0967-4439-a8ef-46b2d1aeccb5" />
<img width="2032" height="1167" alt="Screenshot 2025-12-03 at 12 03 23 PM" src="https://github.com/user-attachments/assets/72149b1c-816c-4191-847a-420accc24521" />
<img width="2032" height="1167" alt="Screenshot 2025-12-03 at 12 03 28 PM" src="https://github.com/user-attachments/assets/74ab1331-ad2f-497a-8950-793cf8822855" />

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [ ] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [ ] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.
- [ ] Bump major, minor, patch, pre-release, and/or build versions [as appropriate](https://semver.org/#semantic-versioning-specification-semver) via the `bump_version` script *if* this repository is versioned *and* the changes in this PR [warrant a version bump](https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api).
- [ ] Create a pre-release (necessary if and only if the pre-release version was bumped).

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Create a release (necessary if and only if the version was bumped).

hawkishpolicy committed:
Merge remote-tracking branch 'origin/Refactor-Filter-Reset-Button-CRASM-3428' into Save-Search-Improvements-CRASM-3418-3429

hawkishpolicy committed:
Refactor SaveSearchModal:

- imported useAreFiltersDefault hook and initialFilters prop to determine if filters are default.
- if filters are default, disbale the " + Save New" button in Drawer Interior to prevent users from saving redundant searches.
- removed space bar from textfield onKeyDown handlers:
  - this was preventing users from entering spaces in the saved search name input.
- "Save" button still uses both Enter and Space to trigger save action for accessibility concerns.

hawkishpolicy committed:
Clean up code.

hawkishpolicy committed:
Build initial unit test for Save Search

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Save-Search-Improvements-CRASM-3418-3429

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Save-Search-Improvements-CRASM-3418-3429

hawkishpolicy committed:
Add the following:

- Reset button now clears active search ID and toggle.
- Refined tests for displaying Save New and Update buttons based on active search state.
- Updated layout test to mock FilterDrawerV2 component.
  - needed because of the addition of useSavedSearchContext import in FilterDrawerV2.tsx
- updated snapshots accordingly.

hawkishpolicy committed:
Draft filter access validation

hawkishpolicy committed:
Refactored create_saved_search call:

- Added current_user parameter to create_saved_search function call in views.py.
- Built a validate_filter_access function to check permissions for standard users.
  - Returns a 403 Forbidden response if standard user tries to save filters for restricted region or organization filters.
- Built an enforce_standard_filters function to modify filters for standard users.
  - Injects region and organization filters to only those the user has access to.

hawkishpolicy committed:
Create pytest case for validating Save Search

hawkishpolicy committed:
Add additional logic for save search calls:

- Prevent standard users from saving default filters (e.g., org and region)
- Added to create_saved_search and update_saved_search functions

hawkishpolicy committed:
Additional validation for saved search endpoint:

- Prevent admin users from saving searches with default regions 1-10
- Wrote new helper functions to query Organizations for all region IDs and return as a simple list
- Updated saved search tests to cover new validation logic.

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Save-Search-Improvements-CRASM-3418-3429

jsalinasnttdata reviewed (dismissed):
LGTM

hawkishpolicy committed:
Ran test coverage

ameliav reviewed (changes requested):
Ran into an error when trying to save a new filter. I selected 1 organization (no other filter selections) and got a 403 error when saving the filter.

hawkishpolicy committed:
Edit prevent_default_admin_filters:

- added emptiness check to org filters
- allows admin users to save searches with all regions if they have at least one org filter

hawkishpolicy committed:
Add check for non region/org filters:

- Allows saving searches with filters that do not include org filters but have all region filters selected.

hawkishpolicy committed:
Add non region/org filter check for standard user:

- Allows standard users to save searches with default region/org filters if other filters are present.
- Prevents saving searches that only contain default region/org filters.

hawkishpolicy commented:
> Ran into an error when trying to save a new filter. I selected 1 organization (no other filter selections) and got a 403 error when saving the filter.

Resolved via: [f18eef7](https://github.com/cisagov/XFD/pull/1373/commits/f18eef7ca05f58e39da9662200be33cd33d57521)
[475aac9](https://github.com/cisagov/XFD/pull/1373/commits/475aac97feed833aa440b44d6681e68e78a3ddb5)
[701ff07](https://github.com/cisagov/XFD/pull/1373/commits/701ff0764368e1d2db81c7acce50aa97f7940267)

ameliav reviewed (approved):
Re-tested and LGTM!
